TESTS/nfc/eeprom/main.cpp

```diff
@@ -0,0 +1,439 @@
+/*
+ * Copyright (c) 2017, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mbed.h"
+#include "greentea-client/test_env.h"
+#include "unity.h"
+#include "utest.h"
+
+#include <events/mbed_events.h>
+#include "NFCEEPROMDriver.h"
+
+using namespace utest::v1;
+
+/* individual steps that map to specific operations that include parameters */
+typedef enum {
+   START_SESSION        = 0x0000,
+
+   END_SESSION          = 0x0100,
+
+   READ_BYTES           = 0x0200,
+   READ_2_BYTES         = 0x0201,
+   READ_4_BYTES         = 0x0202,
+   READ_4_BYTES_MIXED   = 0x0203,
+   READ_4_BYTES_OFFSET  = 0x0204,
+   READ_4_BYTES_ERASED  = 0x0205,
+   READ_4_BYTES_FAIL    = 0x0206,
+
+   WRITE_BYTES          = 0x0300,
+   WRITE_2_BYTES        = 0x0301,
+   WRITE_2_BYTES_OFFSET = 0x0302,
+   WRITE_4_BYTES        = 0x0303,
+   WRITE_4_BYTES_FAIL   = 0x0304,
+
+   ERASE_BYTES          = 0x0400,
+   ERASE_4_BYTES        = 0x0401,
+
+   READ_SIZE            = 0x0500,
+   READ_SIZE_2          = 0x0501,
+   READ_SIZE_4          = 0x0502,
+
+   WRITE_SIZE           = 0x0600,
+   WRITE_SIZE_2         = 0x0601,
+   WRITE_SIZE_4         = 0x0602,
+
+   TERMINATE            = 0xFFFF
+} TestCommand_t;
+
+/* We group the command based on their fist byte to simplify step checking.
+ * Individual conditions of a step are checked in the event so this doesn't
+ * sacrifice any correctness checking */
+const size_t TEST_COMMAND_GROUP_MASK = 0xFF00;
+
+/* test case sequences that index into the array of commands which comprise them */
+typedef enum {
+   SESSION_TEST,
+   WRITE_READ_TEST,
+   ERASE_TEST,
+   WRITE_READ_SIZE_TEST,
+   TRUNCATE_TEST,
+   WRITE_TRUNCATE_TEST,
+   WRITE_WITH_OFFSET_TEST,
+   SEQUENCE_MAX
+} TestSequence_t;
+
+static EventQueue event_queue(/* event count */ 10 * EVENTS_EVENT_SIZE);
+
+static const size_t MAX_STEP = 10;
+static const size_t BUFFER_MAX = 10;
+
+/* defines the "script" the test follows, each step is checked and all
+ * have to execute to completion for a successful run */
+static const TestCommand_t SEQUENCES[SEQUENCE_MAX][MAX_STEP] = {
+    /* SESSION_TEST */
+    { START_SESSION, END_SESSION, TERMINATE },
+
+    /* WRITE_READ_TEST */
+    { START_SESSION, WRITE_SIZE_4, WRITE_4_BYTES, READ_4_BYTES, WRITE_2_BYTES,
+      READ_4_BYTES_MIXED, END_SESSION, TERMINATE },
```

Comment:
Here and in the erase test, you write the size but don't read it back

Reply:
yes, size changes are tested elsewhere, this size write is part of the writing of bytes which is tested by reading them back, no harm checking the size though